helix-core/src/movement.rs

```diff
@@ -1,184 +1,254 @@
+}
 
-        with_end = skip_over_prev(slice, &mut end, char::is_whitespace);
+// ---- util ------------
 
-        // refetch
-        let ch = slice.char(end);
+/// Private helpers to help manipulate slice indices
+pub trait SliceIndexHelpers {
+    fn outside(&self, slice: RopeSlice) -> bool;
+    fn inside(&self, slice: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_boundary(&self, slice: RopeSlice) -> bool;
+    fn category(&self, slice: RopeSlice) -> Option<Category>;
+    /// Returns the start of a word/punctuation group followed by any amount of whitespace.
+    fn start_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group followed by any amount of whitespace.
+    fn end_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group.
+    fn end_of_word(&self, slice: RopeSlice) -> Self;
+    fn skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn backwards_skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn is_whitespace(&self, slice: RopeSlice) -> bool;
+    fn is_end_of_line(&self, slice: RopeSlice) -> bool;
+}
 
-        if is_word(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_punctuation);
-        }
+impl SliceIndexHelpers for usize {
+    fn inside(&self, slice: RopeSlice) -> bool {
+        *self < slice.len_chars()
     }
 
-    Some(Range::new(begin, if with_end { end } else { end + 1 }))
-}
-
-pub fn move_next_word_end(slice: RopeSlice, mut begin: usize, count: usize) -> Option<Range> {
-    let mut end = begin;
+    fn outside(&self, slice: RopeSlice) -> bool {
+        !self.inside(slice)
+    }
 
-    for _ in 0..count {
-        if begin + 2 >= slice.len_chars() {
-            return None;
-        }
+    fn is_boundary(&self, slice: RopeSlice) -> bool {
+        (self + 1).inside(slice)
+            && (categorize(slice.char(*self)) != categorize(slice.char(self + 1)))
+    }
 
-        let ch = slice.char(begin);
-        let next = slice.char(begin + 1);
+    fn category(&self, slice: RopeSlice) -> Option<Category> {
+        self.inside(slice).then(|| categorize(slice.char(*self)))
+    }
 
-        if categorize(ch) != categorize(next) {
-            begin += 1;
-        }
+    fn end_of_word(&self, slice: RopeSlice) -> Self {
+        // Scan the entire slice
+        (*self..slice.len_chars())
+            // Skip any initial newlines, as they must be skipped over for
+            // the purposes of word movement
+            .skip_while(|i| is_end_of_line(slice.char(*i)))
+            // Find the first boundary that doesn't start from whitespace
+            .find(|pos| {
+                pos.is_boundary(slice)
+                    && (!pos.is_whitespace(slice) || (pos + 1).is_end_of_line(slice))
```

Comment:
I think this only needs two `.char` but we did it too many times here. `.char` is `O(log N)` and if rust compiler cannot optimize this means we are getting 2x unnecessary calls.